Parse/src/main/java/com/parse/ParseUser.java

```diff
@@ -501,6 +501,18 @@ private void restoreAnonymity(Map<String, String> anonymousData) {
   }
 
   @Override
+  /* package */ void setState(ParseObject.State newState) {
+    // Avoid overwriting session token
+    State.Builder builder = newState.newBuilder();
+    if (isCurrentUser() && getSessionToken() != null &&
+            newState.get("sessionToken") == null) {
+      builder.put("sessionToken", getSessionToken());
+    }
+
+    super.setState(builder.build());
+  }
```

Comment:
It seems like we'd be creating unnecessary `State` objects with every `setState()`. Mind only creating a new `State` if necessary?